### ISSUE TITLE
[Question] More than 16 physical output?

### ISSUE THREAD

**Issue by guiray** [EnvelopSound/EnvelopForLive#25]
I am trying to figure out how to have more than 16 physical output with this. From the "custom" decoder side it's fine but i'm wondering how the master bus is handling this. Do I need to modify the master bus as well? Is this limited to 16 output?

Just as info I want to try EnvelopForLive in a 31 channel speaker dôme.

**Comment by mcslee** [comment]
![screen shot 2018-04-20 at 2 23 05 pm](https://user-images.githubusercontent.com/597850/39074526-9f729e48-44a6-11e8-9fac-5ef0650b014a.png)

**Comment by mbncr** [comment]
Hi @mcslee 
Thanks for the detailed process, and thanks @guiray for bringing that up ! 
I'm almost there, and really close to lock that 31 channel setup.

I'm facing two problems right now : 
1) Can't get my 16th 'e4l.live.routing.channel_selector.maxpat' to work (screenshot below). I am stuck at 30 outputs. My guess is that one pair being reserved for Binaural, this brings me to 17 pairs of outputs, Ableton won't let me exceed 16 pairs. Would that be correct?
![screen shot 2019-03-07 at 10 19 26 am](https://user-images.githubusercontent.com/31742493/53967686-ab678780-40c3-11e9-8874-f6a4f76cb8d4.png)

2) What would be the best way to bring this customisation to the .amxd pack, without having to add files to Documents/Max8/Max for Live patchers/E4L... 

Thanks in advance.

PS : I thought it was a bit too custom to reopen it as an issue, please change it it you think it makes more sense.

**Comment by mcslee** [comment]
> this brings me to 17 pairs of outputs, Ableton won't let me exceed 16 pairs. Would that be correct?

That's right, Ableton has max 32 channels per in/out on the Max for Live devices. We use the first two outputs to route onto the normal ableton track bus. They're not actually reserved for the Binaural decoder in particular. They're just reserved for routing thru the track bus.

However, since you're modifying this thing, you are welcome to change that and just hijack the first two outputs for your own decoder outputs.

This is the relevant spot:
<img width="470" alt="screen shot 2019-03-07 at 1 17 35 pm" src="https://user-images.githubusercontent.com/597850/53989691-64858c00-40db-11e9-8c27-a01c65cd634f.png">

See how the decoder routes its outputs to outputs 3+, and there is a gate to control whether the first two outputs are also duplicated to outputs 1+2.

You could just rewire this so that the decoder uses up all 32 outputs.

> What would be the best way to bring this customisation to the .amxd pack, without having to add files to Documents/Max8/Max for Live patchers/E4L...

I'm not sure if I totally understand the question. The best thing to do is to make a modified version of the E4L Master Bus patch with a different name. Then you can use the "Freeze" functionality of Max for Live (look for the little snowflake icon in the bottom left) when you're done editing. That will package your device up into a single AMXD file that contains all of the necessary dependencies. You should be able to load it from anywhere at that point.

You may need to click the "Show Containing Project" icon in the bottom tray to make sure that your decoder patcher is listed as a dependency, unless you have hardcoded it into your modified device.

Since you're making modifications to change the outputs already, I might recommend you just get rid of the decoder drop-down and hard-wire in your own decoder patch here. It'll save some complexity.

**Comment by mbncr** [comment]
Thank you @mcslee, precisely what I needed.
I will start from there.

**Comment by bavovk** [comment]
Just as a notice, it seems > 32 channels is possible for M4L:
[https://www.youtube.com/watch?v=nea2bKMgYG0](url)

**Comment by bavovk** [comment]
Just FYI, every Ableton 10 track can now output at least 64 channels as opposed to the 32 mentioned above (demonstrated for 10 Beta [https://www.youtube.com/watch?v=nea2bKMgYG0](url) ). 
It seems 64 is also the maximum for the plugout~ object in Max in my own tests. 

I am modifying the encoder and decoder in respectively the SourcePanner and MasterBus to deal with 64 channels and it seems to work on first glance. Happy to hear any comments or if I should modify e4l.bus.send or e4l.lx.register as well.

**Comment by mcslee** [comment]
Good to know that's a possibility. I don't think it makes sense to change bus.send or lx.register etc. We are sending 3rd-order Ambisonics between E4L devices, which uses 16-channels. There's no need for additional channels when sending Ambisonics, the additional channels would just go unused and consume CPU resources.

For the same reason, there's no reason for the source panner to have additional outputs. It only needs 18 channels - 2 to pass stereo audio thru the normal track chain + 16 to send the 3rd-order ambisonics signal to the master bus.

Per the earlier convo on the thread, the E4L Master Bus can be modified to have more outputs if you want to directly decode to more than 16 speakers within Ableton Live. We will leave this as a custom modification though, since it's not a common use case and we generally recommend using a separate machine / process for decoding when getting into the realms of such large speaker counts.

**Comment by mbncr** [comment]
Hey !
Just pinging back to say that I've managed to have a stable 31 output setup, I'll have more opportunities to test it deeper in the weeks to come.
